python/hpsmc/alignment/_apply.py

```diff
@@ -0,0 +1,306 @@
+    Required Parameters:
+    - **detector**: name of detector to apply parameters to
+
+    Optional Parameters:
+    - **res\_file**: path to millepede results file (default: 'millepede.res')
+    - **bump**: generate the next detector name by incrementing the iter number of the input detector (default: True)
+    - **force**: override the next detector path (default: False)
+    - **next\_detector**: provide name of next detector, preferred over **bump** if provided (default: None)
+    """
+
+    logger = logging.getLogger('ApplyPedeRes')
+
+    def __init__(self) :
+        # optional job
+        self.res_file = 'millepede.res'
+        self.bump = True
+
+        # hidden job parameters
+        self.to_float = 'UNKNOWN'
+
+        super().__init__('ApplyPedeRes')
+
+    def optional_parameters(self) :
+        return super().optional_parameters().extend(['res_file','bump','to_float'])
+
+    def execute(self, log_out, log_err) :
+        if self.bump or self.next_detector is not None :
+            ApplyPedeRes.logger.info('Creating new detector directory.')
+            # deduce source directory and check that it exists
+            src_path = self._detector_dir(self.detector)
+            if not os.path.isdir(src_path) :
+                raise ValueError(f'Detector {self.detector} is not in hps-java')
+            
+            if self.next_detector is None :
+                ApplyPedeRes.logger.info('Deducing next detector name from current name')
+                # deduce iter value, using iter0 if there is no iter suffix
+                matches = re.search('.*iter([0-9]*)', self.detector)
+                if matches is None :
+                    raise ValueError('No "_iterN" suffix on detector name.')
+                else :
+                    i = int(matches.group(1))
+                    self.next_detector = self.detector.replace(f'_iter{i}',f'_iter{i+1}')
+
+            ApplyPedeRes.logger.info(f'Creating new detector named "{self.next_detector}"')
+    
+            # deduce destination path, and make sure it does not exist
+            dest_path = self._detector_dir(self.next_detector)
+            if os.path.isdir(dest_path) and not self.force :
+                raise ValueError(f'Detector {self.next_detector} already exists and so it cannot be created. Use "force" to overwrite an existing detector.')
+                return 3
+    
+            # make copy
+            shutil.copytree(src_path, dest_path, dirs_exist_ok = True)
+    
+        # now we have bumped or not, so reconstruct detector path and check that it exists
+        path = self._detector_dir(self.next_detector)
+        if not os.path.isdir(path) :
+            raise ValueError(f'Detector {self.next_detector} is not in hps-java')
+    
+        # get list of parameters and their MP values
+        parameters = Parameter.parse_pede_res(self.res_file, skip_nonfloat=True)
+
+        self._to_compact(parameters, self.next_detector, save_prev = (not self.bump))
```

Comment:
Here I use `next_detector` which may not be `detector` if `bump` is True or `next_detector` is provided